### ISSUE TITLE
[Question] - DbContext in seperate library project

### ISSUE THREAD

**Issue by dazinator** [dotnet/efcore#6514]
### Steps to reproduce

Create a new asp.net core website project.
Add a new library project to the same solution. Add a new DbContext to the library project, with a basic model.
The question is how to use the `dotnet ef` command line, to work with a DbContext that is in a library project referenced by the main application.
### The issue

In previous versions, I could do this using a startup project argument, and a db context name argument. 
Now using the 1.0.0-preview2-final tooling, I can't seem to get the arguments right no matter what I try.

If i type `dotnet ef --help` this is the help screen I get:

```
Entity Framework .NET Core CLI Commands 1.0.0-preview2-21431
Usage: dotnet ef [options] [command]
Options:
  -h|--help                      Show help information
  -v|--verbose                   Enable verbose output
  --version                      Show version information
  --assembly <ASSEMBLY>          The assembly file to load.
  --startup-assembly <ASSEMBLY>  The assembly file containing the startup class.

  --data-dir <DIR>               The folder used as the data directory (defaults
 to current working directory).
  --project-dir <DIR>            The folder used as the project directory (defau
lts to current working directory).
  --content-root-path <DIR>      The folder used as the content root path for th
e application (defaults to application base directory).
  --root-namespace <NAMESPACE>   The root namespace of the target project (defau
lts to the project assembly name).
Commands:
  database    Commands to manage your database
  dbcontext   Commands to manage your DbContext types
  migrations  Commands to manage your migrations
Use "dotnet ef [command] --help" for more information about a command.
```

It's also not clear to me via the above help screen:-
1. Which directory should I open the command line from - the library project or the web application? Or will it work from either as long as the tool package is in the project.json?
2. What is the "--assembly" argument for. The description says: "The assembly file to load." - this is not helpful. To load for what purpose? (i.e to find the DbContext? Or the Startup Class?)
3. Whats the data-dir used for?

I can't find  any docs for this scenario either.

**Comment by SimonOrdo** [comment]
Are you getting any errors?  

In Startup.cs, have you tried something like...
`services.AddDbContext<ApplicationDbContext>(options =>options.UseSqlServer(connection, b => {b.MigrationsAssembly("YourLibraryProject")` }));`

**Comment by Muchiachio** [comment]
I would like to find a better way for doing this too, but for now I'm using [this](https://github.com/NonFactors/MVC6.Template/blob/develop/src/MvcTemplate.Data/Startup.cs#L12) solution.
Which works with main application configuration, but commands need to be run from library project folder.

**Comment by dazinator** [comment]
@SimonOrdo - it doesn't look like its getting as far as running the application entry point. 

Here is the main entry point for the Web Application, I am simply throwing an exception:

```
 public static void Main(string[] args)
 {
            throw new Exception("I ran!");
 }
```

Then from within the wen application project directory I am running:

`dotnet ef migrations list -c MigrationDbContext --verbose`

Where `MigrationDbContext` is the name of the DbContext class in a project that is referenced by the web application.

This is the full verbose output I get:

```

D:\Repos\gluon2\src\Gluon.WebApplication>dotnet ef migrations list -c MigrationD
bContext --verbose
Using target project 'Gluon.WebApplication'.
Using startup project 'Gluon.WebApplication'.
Using framework 'net452'.
Using configuration 'Debug'.
Running C:\Program Files\dotnet\dotnet.exe build D:\Repos\gluon2\src\Gluon.WebAp
plication\project.json --configuration Debug --framework net452
Process ID: 14796
Project Gluon.Core (.NETFramework,Version=v4.5.2) was previously compiled. Skipp
ing compilation.
Project Smidge (.NETFramework,Version=v4.5.2) was previously compiled. Skipping
compilation.
Project Gluon.Module.Bundler (.NETFramework,Version=v4.5.2) was previously compi
led. Skipping compilation.
Project Gluon.Module.Amd (.NETFramework,Version=v4.5.2) was previously compiled.
 Skipping compilation.
Project Gluon.Module.Migrator (.NETFramework,Version=v4.5.2) was previously comp
iled. Skipping compilation.
Project Gluon.Module.NavMenu (.NETFramework,Version=v4.5.2) was previously compi
led. Skipping compilation.
Project Gluon.Module.PackageManager (.NETFramework,Version=v4.5.2) was previousl
y compiled. Skipping compilation.
Project Gluon.Module.Serilog (.NETFramework,Version=v4.5.2) was previously compi
led. Skipping compilation.
Project Gluon.Module.PubSubJs (.NETFramework,Version=v4.5.2) was previously comp
iled. Skipping compilation.
Project Gluon.Module.Environment (.NETFramework,Version=v4.5.2) was previously c
ompiled. Skipping compilation.
Project Gluon.Module.CookieContents (.NETFramework,Version=v4.5.2) was previousl
y compiled. Skipping compilation.
Project Gluon.Module.ThemeSwitcher (.NETFramework,Version=v4.5.2) was previously
 compiled. Skipping compilation.
Project Gluon.Module.JsnLog (.NETFramework,Version=v4.5.2) was previously compil
ed. Skipping compilation.
Project Gluon.Module.UserVoice (.NETFramework,Version=v4.5.2) was previously com
piled. Skipping compilation.
Project Gluon.Module.DataProtection (.NETFramework,Version=v4.5.2) was previousl
y compiled. Skipping compilation.
Project Gluon.Module.Smtp (.NETFramework,Version=v4.5.2) was previously compiled
. Skipping compilation.
Project Gluon.WebApplication (.NETFramework,Version=v4.5.2) was previously compi
led. Skipping compilation.
Setting the data directory to 'D:\Repos\gluon2\src\Gluon.WebApplication\bin\Debu
g\net452\win7-x64\'.
Invoking dependency command 'Microsoft.EntityFrameworkCore.Design' in project 'G
luon.WebApplication'
Running D:\Repos\gluon2\src\Gluon.WebApplication\bin\Debug\net452\win7-x64\Micro
soft.EntityFrameworkCore.Design.exe --assembly D:\Repos\gluon2\src\Gluon.WebAppl
ication\bin\Debug\net452\win7-x64\Gluon.WebApplication.exe --startup-assembly D:
\Repos\gluon2\src\Gluon.WebApplication\bin\Debug\net452\win7-x64\Gluon.WebApplic
ation.exe --dispatcher-version 1.0.0-preview2-21431 --data-dir D:\Repos\gluon2\s
rc\Gluon.WebApplication\bin\Debug\net452\win7-x64\ --project-dir D:\Repos\gluon2
\src\Gluon.WebApplication --content-root-path D:\Repos\gluon2\src\Gluon.WebAppli
cation --root-namespace Gluon.WebApplication --verbose migrations list -c Migrat
ionDbContext
Process ID: 8332
Finding DbContext classes...
No DbContext named 'MigrationDbContext' was found.

```

So it doesn't find the `DbContext` class. But it doesn't encounter the exception either. 
So is the entry point now irrelevant?

**Comment by dazinator** [comment]
Ok, solved this by doing the following.

Add this class to your library project that has the DbContext:

```
     
    public class ToGetEfMigrationsToWork
    {
        public class TemporaryDbContextFactory : IDbContextFactory<YourDbContext>
        {
            public YourDbContext Create(DbContextFactoryOptions options)
            {
                // Advice is to hard code connection string here.  This only used at dev time
                var builder = new DbContextOptionsBuilder<MenuContext>();
                builder.UseSqlServer("Server=(localdb)\\mssqllocaldb;Database=devdb;Trusted_Connection=True;MultipleActiveResultSets=true");
                return new YourDbContext(builder.Options);
            }
        }
    }

```

open a command prompt within your library project:

```
dotnet ef --startup-project ../YourWebApplication migrations add InitialCreate -c YourDbContext

```

**Comment by bricelam** [comment]
Looks like everything here is working. Closing as a dupe of #5320.

**Comment by bricelam** [comment]
P.S. Migrations will never call your app entry point--that could start a web server or cause other negative side effects at design time.

**Comment by dazinator** [comment]
@bricelam - Thanks for following up with this. In the end, to get things working, I found it necessary to add a reference from my main application (i.e the one with the entry point) to my library projject containing the db project. Then I had to execute the `ef` commands with a startup project specifying my main application. Furthermore, I had to implement a Factory class for the DbConext so that EF could discover and create the DbContext (From memory I want to say `DbContextFactory<TContext>`). 

> Migrations will never call your app entry point

It did seem to insist on running against the main application (i.e the project with an entry point) which is why I had to specify my main application as the startup project. The alternative was that I make my library containing the DbContext have an entry point which meant making it turn into an exe! No thanks!  Perhaps I am doing something wrong.. but happy that #5320 is the issue that tracks this.
